### PR TITLE
network/dhcp: allow to specify custom duid type

### LINKED ISSUE #28897: network,dhcp: DUIDType=custom PR needs follow-ups
### systemd version the issue has been seen with

HEAD

### Used distribution

_No response_

### Linux kernel version used

_No response_

### CPU architectures issue was seen on

None

### Component

_No response_

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

This is reminder.
See post-merge review for https://github.com/systemd/systemd/pull/28846.

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Follow-up for #28846.
Closes #28897.

### CODE REVIEW

File: src/libsystemd-network/dhcp-identifier.h [hunk 1]
```diff
@@ -40,7 +40,7 @@ struct duid {
                 struct {
                         /* DUID_TYPE_EN */
                         be32_t pen;
-                        uint8_t id[8];
+                        uint8_t id[0];
```
Comment (keszybz) [inline src/libsystemd-network/dhcp-identifier.h]: Can we use `[]` here (and in other fields in this union)?

File: man/networkd.conf.xml [hunk 2]
```diff
@@ -127,7 +127,7 @@
         <ulink url="https://tools.ietf.org/html/rfc3315#section-9">RFC 3315</ulink>
         for a description of all the options.</para>
 
-        <para>The following values are understood:
+        <para>This takes an integer in the range 0…65535, or following string values:
```
Comment (keszybz) [inline man/networkd.conf.xml]: "or one of the following"

File: src/network/networkd-dhcp-common.c [hunk 3]
```diff
@@ -1154,9 +1154,16 @@ int config_parse_duid_type(
...
+                uint16_t t;
+
+                r = safe_atou16(type_string, &t);
+                if (r < 0) {
+                        log_syntax(unit, LOG_WARNING, filename, line, r,
+                                   "Failed to parse DUID type '%s', ignoring.", type_string);
+                        return 0;
+                }
+
+                type = t;
```
Comment (keszybz) [inline src/network/networkd-dhcp-common.c:1166]: Hmm, how do we know that `DUIDType` is wide enough? Maybe the type width needs to be forced:
```C
typedef enum DUIDType {
        DUID_TYPE_LLT       = 1,
        DUID_TYPE_EN        = 2,
        DUID_TYPE_LL        = 3,
        DUID_TYPE_UUID      = 4,
        _DUID_TYPE_MAX,
        _DUID_TYPE_INVALID  = -EINVAL,
        _DUID_TYPE_FORCE_U16 = UINT16_MAX,
} DUIDType;
```
See `_SD_ENUM_FORCE_S64`.

Please also add `assert(type == t)` here.

**Review by keszybz — APPROVED** [review]
LGTM, apart from some small nitpicks.

**Comment by yuwata** [comment]
@keszybz Thank you for the review. The suggested changes are applied. Upgrading the green label.

**Review by keszybz — APPROVED** [review]
LGTM.

**Pull request merged**